### PR TITLE
Uplynk connector Ad events

### PR DESCRIPTION
**Pull request by OlegRyz**

### CODE REVIEW

File: connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/network/UplynkAds.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.theoplayer.android.connector.uplynk.network
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class UplynkAds(
```
Comment (MattiasBuelens) [inline connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/network/UplynkAds.kt:6]: Some more docs. 📝

File: connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/AdHandler.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import com.theoplayer.android.api.ads.AdBreak
+import com.theoplayer.android.api.ads.AdBreakInit
+import com.theoplayer.android.api.ads.AdInit
+import com.theoplayer.android.api.ads.ServerSideAdIntegrationController
+import com.theoplayer.android.connector.uplynk.network.UplynkAd
+import com.theoplayer.android.connector.uplynk.network.UplynkAdBreak
+import java.util.WeakHashMap
+
+private val Float.secToMs: Int
+    get() = (this * 1000).toInt()
```
[the remark this answers is not in the document]
Comment (OlegRyz) [reply connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/AdHandler.kt]: I used duration for all properties that represent time

File: connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/AdHandler.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        adBreak.ads.fold(adBreakInit.timeOffset) { offset, item ->
+            scheduledAds[item] = controller.createAd(
+                AdInit(
+                    type = "linear",
+                    timeOffset = offset,
+                    duration = item.duration.secToMs
+                ), currentAdBreak
+            )
+            offset + item.duration.secToMs
+        }
```
[the remark this answers is not in the document]
Comment (OlegRyz) [reply connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/AdHandler.kt]: Done. Didn't understand this from API

File: connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/AdHandler.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    fun onAdProgressUpdate(currentAd: UplynkAdState, adBreak: UplynkAdBreak, time: Double) {
+        val ad = scheduledAds[currentAd.ad]
+        check(ad != null) { "Cannot find an ad: $currentAd" }
+
+        val playedDuration = adBreak.ads
+            .takeWhile { it != currentAd.ad }
+            .sumOf { it.duration.toDouble() }
+
+        val startTime = adBreak.timeOffset + playedDuration
+        val progress = (time - startTime) / currentAd.ad.duration
```
[the remark this answers is not in the document]
Comment (OlegRyz) [reply connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/AdHandler.kt]: Done.

File: connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/network/PreplayResponse.kt [hunk 5]
```diff
@@ -34,4 +34,6 @@ data class PreplayResponse(
      *  * Possible return value: 'https://content-ause2.uplynk.com/'
      *
      */
-    val prefix: String)
+    val prefix: String,
+
+    val ads: UplynkAds)
```
[the remark this answers is not in the document]
Comment (OlegRyz) [reply connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/network/PreplayResponse.kt:45]: Yes, done. I forgot to add some documentation here and in other places

**Review by hovig-theo — APPROVED** [review]

File: connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/UplynkAdIntegration.kt [hunk 6]
```diff
@@ -15,10 +16,17 @@ internal class UplynkAdIntegration(
     private val uplynkDescriptionConverter: UplynkSsaiDescriptionConverter,
     private val uplynkApi: UplynkApi
 ) : ServerSideAdIntegrationHandler {
-
+    private var adScheduler: UplynkAdScheduler? = null
```
[the remark this answers is not in the document]
Comment (MattiasBuelens) [reply connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/UplynkAdIntegration.kt:21]: Sounds like something we might want to fix in THEOplayer itself.

Anyway, setting it to `null` in both `setSource` and `resetSource` will do for now. 👍

**Review by MattiasBuelens — APPROVED** [review]

**Pull request merged**